lib/active_record/immutable.rb

```diff
@@ -0,0 +1,49 @@
+module ActiveRecord
+  module Immutable
+    extend ActiveSupport::Concern
+
+    FIELD_NAME_IDENTIFIER_IMMUTABLE = 'previous_version_id'.freeze
+
+    included do
+      def self.default_scope
+        where(%{
+          not exists (
+            select 1
+            from #{quoted_table_name} model
+            where model.#{FIELD_NAME_IDENTIFIER_IMMUTABLE} = #{quoted_table_name}.id
+          )
+        })
+      end
+
+      def _update_record(_attribute_names = self.attribute_names)
+        create_new_version
+      end
+
+      def create_new_version
+        clone = build_clone
+        clone.immutable_field_value = id
+        clone.save
+      end
+
+      def build_clone
```

Comment:
I feel we can directly call the `dup` where this method is being called, instead of having this just as an alias. What do you think?

Reply:
Yes no problem. I created this method because it will be overwritten.